sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -110,7 +115,9 @@ export class BatchingReceiver extends MessageReceiver {
       };
 
       // Final action to be performed after maxMessageCount is reached or the maxWaitTime is over.
-      const finalAction = (): void => {
+      const finalAction = (this._finalActionHandler = (): void => {
+        // clear finalActionHandler so that it can't be called multiple times.
+        this._finalActionHandler = undefined;
```

Comment:
I like how this is expected to be called just once and that we are enforcing that expectation - but since we aren't resolving the promise if it's a drain scenario, wouldn't it be a problem?

If I remember correctly, when it's the scenario for drain to occur the promise doesn't get resolved as can be seen at L162 and gets resolved only when the finalAction gets invoked again.

Reply:
Does drain still need to be called if `onDetached` is called? Since we close the link in that case, I assumed drain didn't need to be explicitly called as well.

Reply:
`onDetached` here is called when the connection dies. So, there is no point in attempting to drain credits. Pending credits should be drained on live AMQP links only.

Reply:
About handling of drain event, assumption seems fair in that it's not necessary to drain for `onDetached` case for when link is closed

My concern was more about being able to invoke `finalAction` multiple times because I just recall that happening..
But it was for a specific odd-ish scenario I think, but nevertheless one that had come up before
https://github.com/Azure/azure-sdk-for-js/pull/6265#discussion_r350369129
https://github.com/Azure/azure-sdk-for-js/pull/6338#issuecomment-558801152
were few instances where the `addCredit(1)` was resulting in `onMessage` handler to be invoked once because 1 credit was available and link was still open - and it was one scenario where the handlers were getting invoked twice I think. 

But the case was one message was getting lost due to drain event handling adding one credit and on handling that and when that made count to be same as expected `maxMessageCount`, finalAction was invoked again to resolve with all the messages collected so far.

This seemed to have surfaced due to an interaction between timer and network calls which needed more investigation anyway as it seems unclear if user code could in anyway result in these behaviors - just surfacing this scenario / aspect so it can be considered in future perhaps~

Reply:
And actually I wonder if there was a specific reason why invoking it multiple times needed to be explicitly disallowed? :?
Doesn't seem like a usual thing to do? but I may be wrong

Reply:
It was basically so that if `onDetached` is called after `finalAction` was called, `finalAction` wouldn't be called again.